network/p2p/unicast/rate_limiter_map.go

```diff
@@ -0,0 +1,114 @@
+package unicast
+
+import (
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+type rateLimiterMetadata struct {
+	// limiter the rate limiter
+	limiter *rate.Limiter
+	// lastAccessed the last timestamp when the limiter was used. This is used to cleanup old limiter data
+	lastAccessed time.Time
+	// lastRateLimit the last timestamp this the peer was rate limited.
+	lastRateLimit time.Time
+}
+
+// rateLimiterMap stores a rateLimiterMetadata for each peer in an underlying map.
+type rateLimiterMap struct {
+	mu              sync.Mutex
+	ttl             time.Duration
+	cleanupInterval time.Duration
+	limiters        map[peer.ID]*rateLimiterMetadata
+	done            chan struct{}
+}
+
+func newLimiterMap(ttl, cleanupInterval time.Duration) *rateLimiterMap {
+	return &rateLimiterMap{
+		mu:              sync.Mutex{},
+		limiters:        make(map[peer.ID]*rateLimiterMetadata),
+		ttl:             ttl,
+		cleanupInterval: cleanupInterval,
+		done:            make(chan struct{}),
+	}
+}
+
+// get returns limiter in rateLimiterMap map
+func (r *rateLimiterMap) get(peerID peer.ID) (*rateLimiterMetadata, bool) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if lmtr, ok := r.limiters[peerID]; ok {
+		lmtr.lastAccessed = time.Now()
+		return lmtr, ok
+	}
```

Comment:
Yes because expired in this case means "peer has been inactive and has not sent any messages" it doesn't mean the limiter is now useless, so we don't need to check for expiration here leave it for the cleanup loop.